### PR TITLE
#950: Refactor options line parser v3: Inject swig factory

### PR DESCRIPTION
**Pull request by tomuben**

### CODE REVIEW

File: exaudfclient/base/javacontainer/test/cpp/javacontainer_test.cc [hunk 1]
```diff
@@ -112,3 +113,206 @@ TEST(JavaContainer, quoted_jvm_option) {
...
+        "class JVMOPTION_TEST_WITH_SPACE {\n"
+        "static void run(ExaMetadata exa, ExaIterator ctx) throws Exception {\n\n"
+        "\tctx.emit(\"Success!\");\n"
+         " }\n}\n";
+    EXPECT_EQ(vm.getJavaVMInternalStatus().m_scriptCode, expected_script_code);
+    EXPECT_EQ(vm.getJavaVMInternalStatus().m_exaJarPath, "/exaudf/base/javacontainer/exaudf_deploy.jar");
+    EXPECT_EQ(vm.getJavaVMInternalStatus().m_classpath, "/tmp:/exaudf/base/javacontainer/exaudf_deploy.jar");
+    EXPECT_TRUE(vm.getJavaVMInternalStatus().m_needsCompilation);
+    /*
+     * Note: The option "DEF" is wrong and causes UDF's to crash!
```
Comment (tkilias) [inline exaudfclient/base/javacontainer/test/cpp/javacontainer_test.cc:152]: Do we need this case in this Test Case. Seems Independent of the Import.

File: exaudfclient/base/javacontainer/test/cpp/javacontainer_test.cc [hunk 2]
```diff
@@ -112,3 +113,206 @@ TEST(JavaContainer, quoted_jvm_option) {
...
```
Comment (tkilias) [inline exaudfclient/base/javacontainer/test/cpp/javacontainer_test.cc:198]: Do we need this case in this Test Case. Seems Independent of the Import

File: exaudfclient/base/javacontainer/test/cpp/javacontainer_test.cc [hunk 3]
```diff
@@ -112,3 +113,206 @@ TEST(JavaContainer, quoted_jvm_option) {
...
+     *       The correct option would be '-Dhttp.agent=\"ABC DEF\"'
+     */
+    const std::vector<std::string> expectedJVMOptions = {   "-Dhttp.agent=\"ABC", "DEF\"", "-Xms128m", "-Xmx128m", "-Xss512k",
+                                                            "-XX:ErrorFile=/tmp/hs_err_pid%p.log",
+                                                            "-Djava.class.path=/tmp:/exaudf/base/javacontainer/exaudf_deploy.jar",
+                                                            "-XX:+UseSerialGC" };
+    EXPECT_EQ(vm.getJavaVMInternalStatus().m_jvmOptions, expectedJVMOptions);
+}
+
+TEST(JavaContainer, import_script_with_jvmoption) {
```
Comment (tkilias) [inline exaudfclient/base/javacontainer/test/cpp/javacontainer_test.cc:208]: Oh gosh. The imported scripts are actually parsed for the other options. Good catch and test.

Comment (tkilias) [reply exaudfclient/base/javacontainer/test/cpp/javacontainer_test.cc:208]: Now we need a test where both scripts have Jvm options and probably other options, too

File: exaudfclient/base/javacontainer/test/cpp/javacontainer_test.cc [hunk 4]
```diff
@@ -112,3 +113,206 @@ TEST(JavaContainer, quoted_jvm_option) {
...
+        "class JVMOPTION_TEST_WITH_SPACE {\n"
+        "static void run(ExaMetadata exa, ExaIterator ctx) throws Exception {\n\n"
+        "\tctx.emit(\"Success!\");\n"
+         " }\n}\n";
+    EXPECT_EQ(vm.getJavaVMInternalStatus().m_scriptCode, expected_script_code);
+    EXPECT_EQ(vm.getJavaVMInternalStatus().m_exaJarPath, "/exaudf/base/javacontainer/exaudf_deploy.jar");
+    EXPECT_EQ(vm.getJavaVMInternalStatus().m_classpath, "/tmp:/exaudf/base/javacontainer/exaudf_deploy.jar");
+    EXPECT_TRUE(vm.getJavaVMInternalStatus().m_needsCompilation);
+    /*
+     * Note: The option "DEF" is wrong and causes UDF's to crash!
```
Comment (tkilias) [inline exaudfclient/base/javacontainer/test/cpp/javacontainer_test.cc:243]: Do we need this case in this Test Case. Seems Independent of the Import

File: exaudfclient/base/javacontainer/test/cpp/javacontainer_test.cc [hunk 5]
```diff
@@ -112,3 +113,206 @@ TEST(JavaContainer, quoted_jvm_option) {
...
+     *       The correct option would be '-Dhttp.agent=\"ABC DEF\"'
+     */
+    const std::vector<std::string> expectedJVMOptions = {   "-Dhttp.agent=\"ABC", "DEF\"", "-Xms128m", "-Xmx128m", "-Xss512k",
+                                                            "-XX:ErrorFile=/tmp/hs_err_pid%p.log",
+                                                            "-Djava.class.path=/tmp:/exaudf/base/javacontainer/exaudf_deploy.jar",
+                                                            "-XX:+UseSerialGC" };
+    EXPECT_EQ(vm.getJavaVMInternalStatus().m_jvmOptions, expectedJVMOptions);
+}
+
+TEST(JavaContainer, multiple_import_scripts) {
```
Comment (tkilias) [inline exaudfclient/base/javacontainer/test/cpp/javacontainer_test.cc:253]: I think, we are missing a test that has a recursion depth of 3 or more.

File: exaudfclient/base/javacontainer/test/cpp/javacontainer_test.cc (hunk 1 above)
Comment (tomuben) [reply exaudfclient/base/javacontainer/test/cpp/javacontainer_test.cc:152]: Done

File: exaudfclient/base/javacontainer/test/cpp/javacontainer_test.cc (hunk 2 above)
Comment (tomuben) [reply exaudfclient/base/javacontainer/test/cpp/javacontainer_test.cc:198]: Done

File: exaudfclient/base/javacontainer/test/cpp/javacontainer_test.cc (hunk 5 above)
Comment (tomuben) [reply exaudfclient/base/javacontainer/test/cpp/javacontainer_test.cc:253]: Done

**Review by tkilias — APPROVED** [review]

**Comment by tomuben** [comment]
https://eu-central-1.console.aws.amazon.com/codebuild/home?region=eu-central-1#/builds/build-batch/ScriptLanguagesCodeBuild-YEjTntAM8ZKN:b21c9215-b32a-4292-bc9d-f08633dd058c/view/new
run can be ignored: It failed due to a timeout in security scan

**Pull request merged**